discovery/triton/triton.go

```diff
@@ -0,0 +1,169 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package triton
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/log"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/config"
+	"github.com/prometheus/prometheus/util/httputil"
+	"golang.org/x/net/context"
+)
+
+const (
+	tritonLabel             = model.MetaLabelPrefix + "triton_"
+	tritonLabelMachineId    = tritonLabel + "machine_id"
+	tritonLabelMachineAlias = tritonLabel + "machine_alias"
+	tritonLabelMachineImage = tritonLabel + "machine_image"
+	tritonLabelServerId     = tritonLabel + "server_id"
+	namespace               = "prometheus"
+)
+
+var (
+	refreshFailuresCount = prometheus.NewCounter(
+		prometheus.CounterOpts{
+			Name: "prometheus_sd_triton_refresh_failures_total",
+			Help: "The number of Triton-SD scrape failures.",
+		})
+	refreshDuration = prometheus.NewSummary(
+		prometheus.SummaryOpts{
+			Name: "prometheus_sd_triton_refresh_duration_seconds",
+			Help: "The duration of a Triton-SD refresh in seconds.",
+		})
+)
+
+func init() {
+	prometheus.MustRegister(refreshFailuresCount)
+	prometheus.MustRegister(refreshDuration)
+}
+
+type DiscoveryResponse struct {
+	Containers []struct {
+		ServerUUID  string `json:"server_uuid"`
+		VMAlias     string `json:"vm_alias"`
+		VMImageUUID string `json:"vm_image_uuid"`
+		VMUUID      string `json:"vm_uuid"`
+	} `json:"containers"`
+}
+
+// Discovery periodically performs Triton-SD requests. It implements
+// the TargetProvider interface.
+type Discovery struct {
+	client   *http.Client
+	interval time.Duration
+	logger   log.Logger
+	sdConfig *config.TritonSDConfig
+}
+
+// New returns a new Discovery which periodically refreshes its targets.
+func New(logger log.Logger, conf *config.TritonSDConfig) (*Discovery, error) {
+	tls, err := httputil.NewTLSConfig(conf.TLSConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	transport := &http.Transport{TLSClientConfig: tls}
+	client := &http.Client{Transport: transport}
+
+	return &Discovery{
+		client:   client,
+		interval: time.Duration(conf.RefreshInterval),
+		logger:   logger,
+		sdConfig: conf,
+	}, nil
+}
+
+// Run implements the TargetProvider interface.
+func (d *Discovery) Run(ctx context.Context, ch chan<- []*config.TargetGroup) {
+	defer close(ch)
+
+	ticker := time.NewTicker(d.interval)
+	defer ticker.Stop()
+
+	// Get an initial set right away.
+	tg, err := d.refresh()
+	if err != nil {
+		d.logger.Error(err)
```

Comment:
Similarly, it should ideally be `d.logger.With("err", err).Error("Refreshing targets failed")`. Same a few lines down.